src/messaging/ReliableMessageMgr.cpp

```diff
@@ -267,7 +267,16 @@ CHIP_ERROR ReliableMessageMgr::AddToRetransTable(ReliableMessageContext * rc, Re
     bool added     = false;
     CHIP_ERROR err = CHIP_NO_ERROR;
 
-    VerifyOrDie(rc != nullptr && !rc->IsOccupied());
+    VerifyOrDie(rc != nullptr);
+
+    if (rc->IsOccupied())
+    {
+        // This can happen if we have a misbehaving peer that is not sending
+        // acks with its application-level responses when it should, so we end
```

Comment:
What does "when it should" mean? Piggybacking acks is not a deterministic operation.

Reply:
It actually needs to be.  Specifically, an application-level response to a message needs to carry the piggyback ack for that message, if the message was sent using MRP.  That was the outcome of the discussion in #9796.

But now that you make me think through this again, I think things are still broken even with the changes here.  Consider this sequence of events, all on the same exchange; numbering only applies to app-level messages:
1. A sends message 1 to B
2. B is slow to respond, A does a resend and the resend is delayed in the network.
3. B responds with message 2, which acks message 1.
4. A sends message 3
5. B sends standalone ack to message 3, which is lost
6. The duplicate message from step 3 is delivered and triggers a standalone ack.
7. B responds with message 4, which carries a piggyback ack... for message 1 (!)
8. A tries to send message 5, but still has an un-acked message 3, and things fail.

I will fix this, but it really seems like we need a spec-level issue to sort this out.  And I don't see a good way to write a test for this situation with our current test infrastructure.  :( @turon @msandstedt @kghost

Reply:
Filed https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/4168 on the spec.

Reply:
Thanks.  I'll take my comments to #4168.